File: ra-idp-go/internal/authentication/usecases/session_manager.go
// Session manager: login session の生成・解決・破棄。
// Cookie 名と TTL を 1 箇所に集約する。
package usecases

import (
	"context"
	"net/url"
	"slices"
	"strings"
	"time"

	"ra-idp-go/internal/authentication/domain"
	"ra-idp-go/internal/authentication/ports"
	"ra-idp-go/internal/shared/spec"
	"ra-idp-go/internal/tenancy"
)

const (
	SessionCookie     = "ra_idp_session"
	SessionTTLSeconds = 3600
)

type SessionManager struct {
	Store ports.SessionStore
}

func NewSessionManager(s ports.SessionStore) *SessionManager {
	return &SessionManager{Store: s}
}

func (m *SessionManager) Create(ctx context.Context, sub string, amr []string, now time.Time) (*domain.AuthenticationContext, error) {
	return m.CreateWithPending(ctx, sub, amr, now, false)
}

func (m *SessionManager) CreateWithPending(
	ctx context.Context,
	sub string,
	amr []string,
	now time.Time,
	authenticationPending bool,
) (*domain.AuthenticationContext, error) {
	if now.IsZero() {
		now = time.Now().UTC()
	}
	id, err := spec.NewUUIDv4()
	if err != nil {
		return nil, err
	}
	sess := &spec.LoginSession{
		ID:                    id,
		TenantID:              tenancy.TenantID(ctx),
		Sub:                   sub,
		AuthTime:              now.Unix(),
		AMR:                   amr,
		ACR:                   DeriveACR(amr),
		AuthenticationPending: authenticationPending,
		ExpiresAt:             now.Add(SessionTTLSeconds * time.Second),
	}
	if err := m.Store.Save(ctx, sess); err != nil {
		return nil, err
	}
	return &domain.AuthenticationContext{
		Sub:                   sub,
		AuthTime:              sess.AuthTime,
		AMR:                   amr,
		ACR:                   sess.ACR,
		SessionID:             id,
		AuthenticationPending: sess.AuthenticationPending,
	}, nil
}

func (m *SessionManager) CompleteFactor(
	ctx context.Context,
	sessionID string,
	additionalAMR []string,
) (*domain.AuthenticationContext, error) {
	sess, err := m.Store.Find(ctx, sessionID)
	if err != nil || sess == nil {
		return nil, err
	}
	if sess.TenantID != tenancy.TenantID(ctx) {
		return nil, nil
	}
	merged := slices.Clone(sess.AMR)
	for _, method := range additionalAMR {
		if !slices.Contains(merged, method) {
			merged = append(merged, method)
		}
	}
	sess.AMR = merged
	sess.ACR = DeriveACR(merged)
	sess.AuthenticationPending = false
	if err := m.Store.Save(ctx, sess); err != nil {
		return nil, err
	}
	return &domain.AuthenticationContext{
		Sub:                   sess.Sub,
		AuthTime:              sess.AuthTime,
		AMR:                   slices.Clone(sess.AMR),
		ACR:                   sess.ACR,
		SessionID:             sess.ID,
		AuthenticationPending: sess.AuthenticationPending,
		StepUpAt:              sess.StepUpAt,
	}, nil
}

// RecordStepUp は session に step-up 再認証の成立時刻を刻む (ADR-043)。pending な
// session や別テナントの session には作用させない。成立後の AuthenticationContext を返す。
func (m *SessionManager) RecordStepUp(
	ctx context.Context,
	sessionID string,
	now time.Time,
) (*domain.AuthenticationContext, error) {
	if now.IsZero() {
		now = time.Now().UTC()
	}
	sess, err := m.Store.Find(ctx, sessionID)
	if err != nil || sess == nil {
		return nil, err
	}
	if sess.TenantID != tenancy.TenantID(ctx) || sess.AuthenticationPending {
		return nil, nil
	}
	sess.StepUpAt = now.Unix()
	if err := m.Store.Save(ctx, sess); err != nil {
		return nil, err
	}
	return &domain.AuthenticationContext{
		Sub:                   sess.Sub,
		AuthTime:              sess.AuthTime,
		AMR:                   slices.Clone(sess.AMR),
		ACR:                   sess.ACR,
		SessionID:             sess.ID,
		AuthenticationPending: sess.AuthenticationPending,
		StepUpAt:              sess.StepUpAt,
	}, nil
}

func (m *SessionManager) Resolve(ctx context.Context, headers domain.Headers) (*domain.AuthenticationContext, error) {
	sid := parseCookies(headers.Get("Cookie"))[SessionCookie]
	if sid == "" {
		return nil, nil
	}
	sess, err := m.Store.Find(ctx, sid)
	if err != nil {
		return nil, err
	}
	if sess == nil {
		return nil, nil
	}
	if sess.TenantID != tenancy.TenantID(ctx) {
		return nil, nil
	}
	return &domain.AuthenticationContext{
		Sub:                   sess.Sub,
		AuthTime:              sess.AuthTime,
		AMR:                   sess.AMR,
		ACR:                   sess.ACR,
		SessionID:             sess.ID,
		AuthenticationPending: sess.AuthenticationPending,
		StepUpAt:              sess.StepUpAt,
	}, nil
}

func (m *SessionManager) Revoke(ctx context.Context, cookieHeader string) error {
	sid := parseCookies(cookieHeader)[SessionCookie]
	if sid == "" {
		return nil
	}
	return m.Store.Delete(ctx, sid)
}

func parseCookies(header string) map[string]string {
	out := map[string]string{}
	if header == "" {
		return out
	}
	for _, part := range strings.Split(header, ";") {
		part = strings.TrimSpace(part)
		name, value, ok := strings.Cut(part, "=")
		if !ok || name == "" {
			continue
		}
		if dec, err := url.QueryUnescape(value); err == nil {
			out[name] = dec
		} else {
			out[name] = value
		}
	}
	return out
}

File: ra-idp/src/authentication/usecases/session-manager.ts
import { randomUUID } from 'crypto'
import type { LoginSession, SessionStore } from '../ports/session-store'
import type {
  AuthenticationContext,
  AuthenticationContextResolver,
} from '../domain/authentication-context'
import type { UserRepository } from '../ports/user-repository'
import { deriveAcr } from './acr-vocabulary'

export const SESSION_COOKIE = 'ra_idp_session'
export const SESSION_TTL_SECONDS = 3600

export interface SessionManager extends AuthenticationContextResolver {
  create(
    tenantId: string,
    sub: string,
    amr: string[],
    now?: Date,
    options?: { authenticationPending?: boolean },
  ): Promise<AuthenticationContext>
  /**
   * 中間セッションに追加 factor を記録し認証完了状態に昇格する。
   * amr の重複は除き、acr は更新後の amr から再導出する。
   */
  completeFactor(sessionId: string, additionalAmr: string[]): Promise<AuthenticationContext | null>
  revoke(cookieHeader: string | undefined): Promise<void>
}

export class LoginSessionManager implements SessionManager {
  /**
   * ADR-031: `userRepo` を渡すと `resolve` が user の `disabled_at` を見て、
   * 無効化されたユーザーの session は副作用付きで失効させ null を返す。
   * userRepo を渡さない場合 (旧テスト互換) は session 単独の検証のみ行う。
   */
  constructor(
    private readonly sessionStore: SessionStore,
    private readonly userRepo?: UserRepository,
  ) {}

  async create(
    tenantId: string,
    sub: string,
    amr: string[],
    now = new Date(),
    options: { authenticationPending?: boolean } = {},
  ): Promise<AuthenticationContext> {
    const acr = deriveAcr(amr)
    const session: LoginSession = {
      id: randomUUID(),
      tenant_id: tenantId,
      sub,
      auth_time: Math.floor(now.getTime() / 1000),
      amr,
      acr,
      authentication_pending: options.authenticationPending ?? false,
      expires_at: new Date(now.getTime() + SESSION_TTL_SECONDS * 1000).toISOString(),
    }
    await this.sessionStore.save(session)
    return {
      sub,
      auth_time: session.auth_time,
      amr,
      acr,
      session_id: session.id,
      authentication_pending: session.authentication_pending,
    }
  }

  async completeFactor(
    sessionId: string,
    additionalAmr: string[],
  ): Promise<AuthenticationContext | null> {
    const session = await this.sessionStore.find(sessionId)
    if (!session) return null
    const mergedAmr = Array.from(new Set([...session.amr, ...additionalAmr]))
    const updated: LoginSession = {
      ...session,
      amr: mergedAmr,
      acr: deriveAcr(mergedAmr),
      authentication_pending: false,
    }
    await this.sessionStore.save(updated)
    return {
      sub: updated.sub,
      auth_time: updated.auth_time,
      amr: updated.amr,
      acr: updated.acr,
      session_id: updated.id,
      authentication_pending: false,
    }
  }

  async resolve(headers: Headers): Promise<AuthenticationContext | null> {
    const sessionId = parseCookies(headers.get('cookie') ?? undefined)[SESSION_COOKIE]
    if (!sessionId) return null

    const session = await this.sessionStore.find(sessionId)
    if (!session) return null
    if (this.userRepo) {
      const user = await this.userRepo.findBySub(session.sub)
      if (!user || user.disabled_at) {
        // ADR-031: 無効化された user の既存セッションは利用できない。
        // 副作用として session を即時失効させる。
        await this.sessionStore.delete(session.id)
        return null
      }
    }
    return {
      sub: session.sub,
      auth_time: session.auth_time,
      amr: session.amr,
      acr: session.acr,
      session_id: session.id,
      authentication_pending: session.authentication_pending,
    }
  }

  async revoke(cookieHeader: string | undefined): Promise<void> {
    const sessionId = parseCookies(cookieHeader)[SESSION_COOKIE]
    if (sessionId) await this.sessionStore.delete(sessionId)
  }
}

function parseCookies(header: string | undefined): Record<string, string> {
  const cookies: Record<string, string> = {}
  if (!header) return cookies
  for (const part of header.split(';')) {
    const [name, ...rest] = part.trim().split('=')
    if (!name) continue
    cookies[name] = decodeURIComponent(rest.join('='))
  }
  return cookies
}
